chainextensions/dapps-staking/impl/src/lib.rs

```diff
@@ -0,0 +1,489 @@
+    DappsStakingValueInput,
+};
+use codec::{Decode, Encode};
+use frame_support::traits::{Currency, Get};
+use frame_system::RawOrigin;
+use pallet_contracts::chain_extension::{Environment, Ext, InitState, SysConfig, UncheckedFrom};
+use pallet_dapps_staking::{RewardDestination, WeightInfo};
+use sp_std::marker::PhantomData;
+
+type BalanceOf<T> = <<T as pallet_dapps_staking::Config>::Currency as Currency<
+    <T as frame_system::Config>::AccountId,
+>>::Balance;
+
+enum DappsStakingFunc {
+    CurrentEra,
+    UnbondingPeriod,
+    EraRewards,
+    EraStaked,
+    StakedAmount,
+    StakedAmountOnContract,
+    ReadContractStake,
+    Register,
+    BondAndStake,
+    UnbondAndUnstake,
+    WithdrawUnbonded,
+    ClaimStaker,
+    ClaimDapp,
+    SetRewardDestination,
+    NominationTransfer,
+}
+
+impl TryFrom<u32> for DappsStakingFunc {
+    type Error = DispatchError;
+
+    fn try_from(value: u32) -> Result<Self, Self::Error> {
+        match value {
+            1 => Ok(DappsStakingFunc::CurrentEra),
+            2 => Ok(DappsStakingFunc::UnbondingPeriod),
+            3 => Ok(DappsStakingFunc::EraRewards),
+            4 => Ok(DappsStakingFunc::EraStaked),
+            5 => Ok(DappsStakingFunc::StakedAmount),
+            6 => Ok(DappsStakingFunc::StakedAmountOnContract),
+            7 => Ok(DappsStakingFunc::ReadContractStake),
+            8 => Ok(DappsStakingFunc::Register),
+            9 => Ok(DappsStakingFunc::BondAndStake),
+            10 => Ok(DappsStakingFunc::UnbondAndUnstake),
+            11 => Ok(DappsStakingFunc::WithdrawUnbonded),
+            12 => Ok(DappsStakingFunc::ClaimStaker),
+            13 => Ok(DappsStakingFunc::ClaimDapp),
+            14 => Ok(DappsStakingFunc::SetRewardDestination),
+            15 => Ok(DappsStakingFunc::NominationTransfer),
+            _ => Err(DispatchError::Other(
+                "DappsStakingExtension: Unimplemented func_id",
+            )),
+        }
+    }
+}
+
+// pub struct DappsStakingExtension;
+pub struct DappsStakingExtension<R>(PhantomData<R>);
+
+impl<T: pallet_dapps_staking::Config> ChainExtensionExec<T> for DappsStakingExtension<T> {
+    fn execute_func<E>(func_id: u32, env: Environment<E, InitState>) -> Result<(), DispatchError>
+    where
+        E: Ext<T = T>,
+        <E::T as SysConfig>::AccountId:
+            UncheckedFrom<<E::T as SysConfig>::Hash> + AsRef<[u8]> + From<[u8; 32]>,
+    {
+        let func_id = DappsStakingFunc::try_from(func_id)?;
+        let mut env = env.buf_in_buf_out();
+
+        match func_id {
+            // DappsStaking - read_current_era()
+            DappsStakingFunc::CurrentEra => {
+                let era_index = pallet_dapps_staking::CurrentEra::<T>::get();
+                env.write(&era_index.encode(), false, None).map_err(|_| {
+                    DispatchError::Other(
+                        "[ChainExtension] DappsStakingExtension failed to write result",
```

Comment:
Would be good to include which function failed exactly. Currently all have the same error message which isn't helpful.